slsa/provenance.go

```diff
@@ -0,0 +1,128 @@
+package slsa
+
+import (
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+
+	"github.com/slsa-framework/slsa-github-generator/github"
+)
+
+// GithubHostedActionsBuilderID is the builder ID for Github hosted actions.
+const GithubHostedActionsBuilderID = "https://github.com/Attestations/GitHubHostedActions@v1"
+
+// WorkflowRun contains information about the build run including the builder,
+// build invocation, and parameters.
+type WorkflowRun struct {
+	// Subjects is information about the generated artifacts.
+	Subjects []intoto.Subject
+
+	// BuildType indicates the type of build that was done. More importantly it
+	// also specifies the format of the BuildConfig.
+	BuildType string
+
+	// BuildConfig is metadata about the build.
+	BuildConfig interface{}
+
+	// GithubContext is the context for the workflow run.
+	GithubContext github.WorkflowContext
+}
+
+// WorkflowParametersV1 contains parameters given to the workflow invocation in v1 format.
+type WorkflowParametersV1 struct {
+	// Version is the version of the
+	Version int `json:"version"`
+
+	// EventName is the name of the event that initiated the workflow run.
+	EventName string `json:"event_name,omitempty"`
+
+	// EventPayload is the full event payload.
+	// See: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows
+	EventPayload interface{} `json:"event_payload,omitempty"`
+
+	// RefType is type of ref that triggered the workflow run.
+	RefType string `json:"ref_type,omitempty"`
+
+	// Ref is the ref that triggered the workflow run.
+	Ref string `json:"ref,omitempty"`
+
+	// BaseRef is the base ref or base branch of the pull request in a workflow
+	// run.
+	BaseRef string `json:"base_ref,omitempty"`
+
+	// HeadRef is ref or source branch of the pull request in a workflow run.
+	HeadRef string `json:"head_ref,omitempty"`
+
+	// Actor is the username of the user that initiated the workflow run.
+	Actor string `json:"actor,omitempty"`
+
+	// SHA1 is the commit SHA that triggered the workflow run.
+	SHA1 string `json:"sha1,omitempty"`
+}
+
+var (
+	parametersVersion = 1
+	audience          = "slsa-framework"
+)
+
+// HostedActionsProvenance generates an in-toto provenance statement in the SLSA
+// v0.2 format for a workflow run on a Github actions hosted runner.
+func HostedActionsProvenance(w WorkflowRun) (*intoto.ProvenanceStatement, error) {
+	t, err := github.RequestOIDCToken(audience)
+	if err != nil {
+		return nil, err
+	}
+
+	return &intoto.ProvenanceStatement{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       w.Subjects,
+		},
+		Predicate: slsa.ProvenancePredicate{
+			BuildType: w.BuildType,
+			Builder: slsa.ProvenanceBuilder{
+				ID: fmt.Sprintf("https://github.com/%s", t.JobWorkflowRef),
+			},
+			Invocation: slsa.ProvenanceInvocation{
```

Comment:
Can you also fill out the `metadata` field? https://slsa.dev/provenance/v0.2#metadata

At a minimum, `metadata.completeness.parameters: true`.

Reply:
I'll add `metadata.completeness.parameters` and follow up on other fields in another PR

Added issue #8

Reply:
Added `metadata.completeness.parameters'